File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return dfs(root, {-1001, 1001});
    }

    bool dfs(TreeNode* root, vector<int> range) {
        if (!root) {
            return true;
        }

        if (root->val >= range[1] || root->val <= range[0]) {
            return false;
        }

        return dfs(root->left, {range[0], min(range[1], root->val)}) && dfs(root->right, {max(range[0], root->val), range[1]});
    }
};

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        if root == None:
            return res

        nodeQueue = collections.deque()
        nodeQueue.append(root)

        while nodeQueue:
            qLen = len(nodeQueue)
            for i in range(qLen):
                node = nodeQueue.popleft()

                # make sure that no NULL nodes gets added to queue
                if node.left:
                    nodeQueue.append(node.left)
                if node.right:
                    nodeQueue.append(node.right)

                # last element in level - rightmost element
                if i == (qLen - 1) and node:
                    res.append(node.val)
        
        return res

        